Points Kubeflow's check_status value to kubeflow.enabled lock file

DnPlas opened the pull request:
Creates lock file after all actions in enable.kubeflow.sh
and points check_status in addon-lists.yaml to that file.
This ensures Kubeflow addon is listed as enabled when
running microk8s status

Fixes #2695

ktsakalozos commented:
We may want to backport this PR to the branches where the kubeflow addon is available (1.21, 1.20,....)

DnPlas commented:
Hi @ktsakalozos I've rebased my dev branch and also changed the target branch to 1.21, which is the last supported version for Kubeflow. Note that I went with the approach of creating a lock file instead of just checking for a resource.

ktsakalozos commented:
@DnPlas I think we also need to remove the lock file when we disable the add on in https://github.com/ubuntu/microk8s/blob/1.21/microk8s-resources/actions/disable.kubeflow.sh

DnPlas commented:
> @DnPlas I think we also need to remove the lock file when we disable the add on in https://github.com/ubuntu/microk8s/blob/1.21/microk8s-resources/actions/disable.kubeflow.sh

Ah you are absolutely right. Let me add that.

addyess commented on microk8s-resources/actions/disable.kubeflow.sh:
```diff
@@ -1,10 +1,20 @@
+    """Mark Kubeflow addon as disabled by removing kubeflow.enabled lock file."""
+    try:
+        snapdata_path = os.environ.get("SNAP_DATA")
+        lock_fname = "{}/var/lock/kubeflow.enabled".format(snapdata_path)
+        subprocess.call(['sudo', 'rm', lock_fname])
```
you might consider python's `pathlib` here.  
I can't comment on the `sudo` here asn i don't know its relevance.  It seems odd to elevate to sudo in your code -- what if the user cannot `sudo`?   What if the user can but didn't know they were doing something needing sudo.  

is it ok to disable if it's already disabled?  I don't know in this context what the intent is - `rm missing_file` will raise an exception but the except case here doesn't indicate why.  

```python
from pathlib import Path

lock_fname = Path(snapdata_path) / "var" / "lock" / "kubeflow.enabled"
lock_fname.unlink(missing_ok=True)   
```

DnPlas replied:
> I can't comment on the sudo here as i don't know its relevance. It seems odd to elevate to sudo in your code -- what if the user cannot sudo? What if the user can but didn't know they were doing something needing sudo.

I followed what I thought was the standard, for kata-containers (see [PR#2174](https://github.com/ubuntu/microk8s/pull/2174)) it is done that way. @ktsakalozos may give us his input here.

> is it ok to disable if it's already disabled? I don't know in this context what the intent is - rm $SNAP_DATA/var/lock/kubeflow.enabled

What do you mean by "is it ok to disable if it's already disabled?"?

addyess replied:
can someone run this back-to-back?  What would be expected to occur? wouldn't the `rm $lock_fname` fail throwing a subprocess error?

```bash
microk8s disable kubeflow
microk8s disable kubeflow
```

knkski replied:
Personally, I'd expect a message like this:

https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/actions/enable.kubeflow.sh#L203-L212

i.e.

> Kubeflow has already been disabled.

DnPlas replied:
I agree, running `microk8s disable <addon>` when the addon is already disabled is something we have to take care, though that is not exclusive to the lock file proposed in this PR, it applies to pretty much everything we run in the disable action. 
I see that microk8s handles this problem for other addons in [this](https://github.com/ubuntu/microk8s/blob/master/scripts/wrappers/common/utils.py#L275) check. I also see that kubeflow is excluded, see [L271](https://github.com/ubuntu/microk8s/blob/master/scripts/wrappers/common/utils.py#L271) in the same file. We could either do the check manually inside `disable.kubeflow.sh` or re-use what microk8s already does for other addons. I am open to both solutions, though I'd prefer the latter as it is more scalable.

addyess commented on microk8s-resources/actions/enable.kubeflow.sh:
```diff
@@ -29,6 +29,16 @@ CONNECTIVITY_CHECKS = [
+    """Mark Kubeflow addon as enabled by creating kubeflow.enabled."""
+    try:
+        snapdata_path = os.environ.get("SNAP_DATA")
+        lock_fname = "{}/var/lock/kubeflow.enabled".format(snapdata_path)
+        subprocess.call(['sudo', 'touch', lock_fname])
```
See comment above concerning pathlib.  There's also a ```Path.touch()``` method

addyess reviewed (commented):
Who should be the owner of this file you're creating?  Root?  the snap owner?

addyess reviewed (approved):
i approve of this if there's future work to be performed.

Then, again, i cannot merge for you in this organization

knkski reviewed (approved):
LGTM, though @ktsakalozos should give final approval
